src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/util/IndexUtils.kt

```diff
@@ -0,0 +1,111 @@
+import org.elasticsearch.cluster.metadata.IndexMetaData
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler
+import org.elasticsearch.common.xcontent.NamedXContentRegistry
+import org.elasticsearch.common.xcontent.XContentParser
+import org.elasticsearch.common.xcontent.XContentType
+
+class IndexUtils {
+    companion object {
+        @Suppress("ObjectPropertyNaming")
+        const val _META = "_meta"
+        const val SCHEMA_VERSION = "schema_version"
+        const val NO_SCHEMA_VERSION = 1L
+
+        var indexManagementSchemaVersion: Long
+            private set
+        var indexManagementHistoryVersion: Long
+            private set
+
+        init {
+            indexManagementSchemaVersion = getSchemaVersion(IndexStateManagementIndices.indexStateManagementMappings())
+            indexManagementHistoryVersion = getSchemaVersion(IndexStateManagementIndices.indexStateManagementHistoryMappings())
+        }
+
+        @JvmStatic
+        fun getSchemaVersion(mapping: String): Long {
+            val xcp = XContentType.JSON.xContent().createParser(NamedXContentRegistry.EMPTY,
+                LoggingDeprecationHandler.INSTANCE, mapping)
+
+            while (!xcp.isClosed) {
+                val token = xcp.currentToken()
+                if (token != null && token != XContentParser.Token.END_OBJECT && token != XContentParser.Token.START_OBJECT) {
+                    if (xcp.currentName() != _META) {
+                        xcp.nextToken()
+                        xcp.skipChildren()
+                    } else {
+                        while (xcp.nextToken() != XContentParser.Token.END_OBJECT) {
+                            when (xcp.currentName()) {
+                                SCHEMA_VERSION -> {
+                                    val version = xcp.longValue()
+                                    require(version > -1)
+                                    return version
+                                }
+                                else -> xcp.nextToken()
+                            }
+                        }
+                    }
+                }
+                xcp.nextToken()
+            }
+            return NO_SCHEMA_VERSION
+        }
+
+        @JvmStatic
+        fun shouldUpdateIndex(index: IndexMetaData, mapping: String): Boolean {
+            var oldVersion = NO_SCHEMA_VERSION
+            val newVersion = getSchemaVersion(mapping)
+
+            val indexMapping = index.mapping()?.sourceAsMap()
+            if (indexMapping != null && indexMapping.containsKey(_META) && indexMapping[_META] is HashMap<*, *>) {
+                val metaData = indexMapping[_META] as HashMap<*, *>
+                if (metaData.containsKey(SCHEMA_VERSION)) {
+                    oldVersion = (metaData[SCHEMA_VERSION] as Int).toLong()
+                }
+            }
+            return newVersion > oldVersion
+        }
+
+        @JvmStatic
+        fun updateIndexMapping(
```

Comment:
nit: Should this method just do what its name implies? Just update the mappings, and leave the shouldUpdateIndex check to be called by the caller of this method?

Reply:
That will lead to duplicate code outside for other places to check `shouldUpdateIndex`. I can rename this function to something like `updateIndexMappingIfNeeded` ? But I think `updateIndexMapping` is fine.